File: education/education/page/fees_dashboard/fees_dashboard.py
# Copyright (c) 2024, Frappe Technologies and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate, add_days
import os


@frappe.whitelist()
def get_page_html():
	"""Get HTML content for the fees dashboard page"""
	return frappe.render_template("education/page/fees_dashboard/fees_dashboard.html", {})
    # page_path = os.path.join(
	# 	frappe.get_module_path("education"),
	# 	"education",
	# 	"page",
	# 	"fees_dashboard",
	# 	"fees_dashboard.html"
	# )
	
	# if os.path.exists(page_path):
	# 	with open(page_path, 'r', encoding='utf-8') as f:
	# 		return f.read()
	return None


@frappe.whitelist()
def get_fees_data(filters=None, page_length=20, page_start=0):
	"""Get fees data for dashboard with filters and pagination"""
	
	if isinstance(filters, str):
		import json
		filters = json.loads(filters)
	
	filters = filters or {}
	
	# Build SQL query
	conditions = ["f.docstatus = 1"]
	
	# Date range filter
	if filters.get("from_date"):
		conditions.append(f"f.posting_date >= '{filters['from_date']}'")
	if filters.get("to_date"):
		conditions.append(f"f.posting_date <= '{filters['to_date']}'")
	
	# Receipt uploaded filter
	if filters.get("receipt_uploaded") == "Yes":
		conditions.append("f.receipt_uploaded = 1")
	elif filters.get("receipt_uploaded") == "No":
		conditions.append("(f.receipt_uploaded = 0 OR f.receipt_uploaded IS NULL)")
	
	# Unpaid filter
	if filters.get("unpaid_only") == "Yes":
		conditions.append("f.outstanding_amount > 0")
	
	where_clause = " AND ".join(conditions)
	
	# Get total count
	count_query = f"""
		SELECT COUNT(*) as total
		FROM `tabFees` f
		WHERE {where_clause}
	"""
	total_count = frappe.db.sql(count_query, as_dict=True)[0].total
	
	# Get data with pagination
	query = f"""
		SELECT 
			f.name as fee_number,
			f.student,
			f.student_name,
			s.name as student_number,
			f.mode_of_payment,
			f.receipt_amount,
			f.outstanding_amount,
			f.grand_total,
			f.receipt_payment_date,
			f.receipt_payer_name,
			f.receipt_uploaded,
			f.posting_date,
			f.due_date,
			f.payment_type
		FROM `tabFees` f
		LEFT JOIN `tabStudent` s ON s.name = f.student
		WHERE {where_clause}
		ORDER BY f.posting_date DESC, f.name DESC
		LIMIT {page_length} OFFSET {page_start}
	"""
	
	data = frappe.db.sql(query, as_dict=True)
	
	# Get attachments for fees with receipts
	fee_names = [d.fee_number for d in data if d.receipt_uploaded]
	if fee_names:
		attachments = frappe.get_all(
			"File",
			fields=["name", "file_name", "file_url", "attached_to_name"],
			filters={
				"attached_to_doctype": "Fees",
				"attached_to_name": ["in", fee_names]
			}
		)
		
		# Map attachments to fees
		attachments_map = {}
		for att in attachments:
			if att.attached_to_name not in attachments_map:
				attachments_map[att.attached_to_name] = []
			attachments_map[att.attached_to_name].append({
				"name": att.name,
				"file_name": att.file_name,
				"file_url": att.file_url
			})
		
		# Add attachments to data
		for d in data:
			d.attachments = attachments_map.get(d.fee_number, [])
	return {
		"data": data,
		"total_count": total_count,
		"page_length": page_length,
		"page_start": page_start
	}


@frappe.whitelist()
def get_dashboard_summary(filters=None):
	"""Get summary statistics for dashboard cards"""
	
	if isinstance(filters, str):
		import json
		filters = json.loads(filters)
	
	filters = filters or {}
	
	conditions = ["f.docstatus = 1"]
	
	if filters.get("from_date"):
		conditions.append(f"f.posting_date >= '{filters['from_date']}'")
	if filters.get("to_date"):
		conditions.append(f"f.posting_date <= '{filters['to_date']}'")
	
	if filters.get("receipt_uploaded") == "Yes":
		conditions.append("f.receipt_uploaded = 1")
	elif filters.get("receipt_uploaded") == "No":
		conditions.append("(f.receipt_uploaded = 0 OR f.receipt_uploaded IS NULL)")
	
	if filters.get("unpaid_only") == "Yes":
		conditions.append("f.outstanding_amount > 0")
	
	where_clause = " AND ".join(conditions)
	
	summary_query = f"""
		SELECT 
			COUNT(*) as total_fees,
			SUM(f.grand_total) as total_amount,
			SUM(f.outstanding_amount) as total_outstanding,
			SUM(CASE WHEN f.receipt_uploaded = 1 THEN 1 ELSE 0 END) as fees_with_receipts,
			SUM(CASE WHEN f.outstanding_amount > 0 THEN 1 ELSE 0 END) as unpaid_fees,
			SUM(f.receipt_amount) as total_receipt_amount
		FROM `tabFees` f
		WHERE {where_clause}
	"""
	
	result = frappe.db.sql(summary_query, as_dict=True)[0]
	
	return {
		"total_fees": result.total_fees or 0,
		"total_amount": flt(result.total_amount) or 0,
		"total_outstanding": flt(result.total_outstanding) or 0,
		"fees_with_receipts": result.fees_with_receipts or 0,
		"unpaid_fees": result.unpaid_fees or 0,
		"total_receipt_amount": flt(result.total_receipt_amount) or 0,
		"paid_amount": flt(result.total_amount) - flt(result.total_outstanding) or 0
	}


@frappe.whitelist()
def accept_receipt_and_pay(fee_name, payment_amount):
	"""Accept receipt and register payment"""
	try:
		fee_doc = frappe.get_doc("Fees", fee_name)
		
		if not fee_doc.receipt_uploaded:
			return {"error": _("No receipt uploaded for this fee")}
		
		payment_amount = flt(payment_amount)
		if payment_amount <= 0:
			return {"error": _("Payment amount must be greater than zero")}
		
		# Register payment
		fee_doc.pay_fee(payment_amount)
		
		return {
			"success": True,
			"message": _("Payment registered successfully")
		}
	except Exception as e:
		frappe.log_error(frappe.get_traceback(), "Accept Receipt Error")
		return {"error": str(e)}


@frappe.whitelist()
def reject_receipt(fee_name):
	"""Reject receipt"""
	try:
		fee_doc = frappe.get_doc("Fees", fee_name)
		fee_doc.reject_receipt()
		frappe.db.commit()
		
		return {
			"success": True,
			"message": _("Receipt rejected successfully")
		}
	except Exception as e:
		frappe.log_error(frappe.get_traceback(), "Reject Receipt Error")
		return {"error": str(e)}


@frappe.whitelist()
def get_fee_details(fee_name):
	"""Get detailed fee information"""
	try:
		fee_doc = frappe.get_doc("Fees", fee_name)
		
		# Get fee components
		components = []
		for comp in fee_doc.components:
			components.append({
				"fees_category": comp.fees_category,
				"description": comp.description,
				"fee_rate": comp.fee_rate,
				"discount": comp.discount,
				"amount": comp.amount
			})
		
		# Get attachments
		attachments = frappe.get_all(
			"File",
			fields=["name", "file_name", "file_url"],
			filters={
				"attached_to_doctype": "Fees",
				"attached_to_name": fee_name
			}
		)
		return {
			"fee_number": fee_doc.name,
			"student": fee_doc.student,
			"student_name": fee_doc.student_name,
			"posting_date": fee_doc.posting_date,
			"due_date": fee_doc.due_date,
			"grand_total": fee_doc.grand_total,
			"outstanding_amount": fee_doc.outstanding_amount,
			"paid_amount": fee_doc.grand_total - fee_doc.outstanding_amount,
			"receipt_uploaded": fee_doc.receipt_uploaded,
			"receipt_amount": fee_doc.receipt_amount,
			"receipt_payer_name": fee_doc.receipt_payer_name,
			"receipt_payment_date": fee_doc.receipt_payment_date,
			"payment_type": fee_doc.payment_type,
			"mode_of_payment": fee_doc.payment_type,
			"components": components,
			"attachments": attachments
		}
	except Exception as e:
		frappe.log_error(frappe.get_traceback(), "Get Fee Details Error")
		return {"error": str(e)}

File: education/education/page/fees_dashboard/fees_dashboard.js
frappe.pages['fees-dashboard'].on_page_load = function(wrapper) {
	// rtl
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'الرسوم',
		single_column: true,
		icon: 'fa fa-money',
		direction: 'rtl'
	});

	// Load HTML content from Python method
	frappe.call({
		method: 'education.education.page.fees_dashboard.fees_dashboard.get_page_html',
		callback: function(r) {
			if (r.message) {
				page.main.html(r.message);
				init_dashboard(page);
			} else {
				// Fallback: show error message
				page.main.html('<div class="alert alert-danger">حدث خطأ في تحميل الصفحة</div>');
			}
		
		}
	});
}

function init_dashboard(page) {
	let me = {
		page: page,
		current_page: 0,
		page_length: 20,
		filters: {},
		summary: {}
	};

	// Initialize filters
	me.init_filters = function() {
		// Set default date range (last 30 days)
		let to_date = frappe.datetime.get_today();
		let from_date = frappe.datetime.add_days(to_date, -30);
		
		$('#filter-from-date').val(from_date);
		$('#filter-to-date').val(to_date);
		
		// Apply filters button
		$('#btn-apply-filters').on('click', function() {
			me.apply_filters();
		});
		
		// Clear filters button
		$('#btn-clear-filters').on('click', function() {
			me.clear_filters();
		});
		
		// Enter key on filters
		$('.filter-group input, .filter-group select').on('keypress', function(e) {
			if (e.which === 13) {
				me.apply_filters();
			}
		});
	};

	me.apply_filters = function() {
		me.filters = {
			from_date: $('#filter-from-date').val() || null,
			to_date: $('#filter-to-date').val() || null,
			receipt_uploaded: $('#filter-receipt-uploaded').val() || null,
			unpaid_only: $('#filter-unpaid').val() || null
		};
		
		// Remove null values
		Object.keys(me.filters).forEach(key => {
			if (me.filters[key] === null || me.filters[key] === '') {
				delete me.filters[key];
			}
		});
		
		me.current_page = 0;
		me.load_data();
		me.load_summary();
	};

	me.clear_filters = function() {
		$('#filter-from-date').val('');
		$('#filter-to-date').val('');
		$('#filter-receipt-uploaded').val('');
		$('#filter-unpaid').val('');
		me.filters = {};
		me.current_page = 0;
		me.load_data();
		me.load_summary();
	};

	me.load_summary = function() {
		frappe.call({
			method: 'education.education.page.fees_dashboard.fees_dashboard.get_dashboard_summary',
			args: {
				filters: me.filters
			},
			callback: function(r) {
				if (r.message && !r.message.error) {
					me.summary = r.message;
					me.render_summary();
				}
			}
		});
	};

	me.render_summary = function() {
		let s = me.summary;
		$('#total-fees').text(s.total_fees || 0);
		$('#total-amount').text(format_currency(s.total_amount || 0));
		$('#total-outstanding').text(format_currency(s.total_outstanding || 0));
		$('#fees-with-receipts').text(s.fees_with_receipts || 0);
		$('#unpaid-fees').text(s.unpaid_fees || 0);
	};

	me.load_data = function() {
		let tbody = $('#fees-table-body');
		tbody.html('<tr><td colspan="9" class="text-center"><div class="loading-spinner"><span>جاري التحميل...</span></div></td></tr>');
		
		frappe.call({
			method: 'education.education.page.fees_dashboard.fees_dashboard.get_fees_data',
			args: {
				filters: me.filters,
				page_length: me.page_length,
				page_start: me.current_page * me.page_length
			},
			callback: function(r) {
				if (r.message && !r.message.error) {
					me.render_table(r.message.data || []);
					me.update_pagination(r.message);
				} else {
					tbody.html('<tr><td colspan="9" class="text-center empty-state"><div class="empty-state-message">حدث خطأ في تحميل البيانات</div></td></tr>');
				}
			}
		});
	};

	me.render_table = function(data) {
		let tbody = $('#fees-table-body');
		
		if (data.length === 0) {
			tbody.html('<tr><td colspan="9" class="text-center empty-state"><div class="empty-state-message">لا توجد بيانات</div></td></tr>');
			return;
		}
		
		let html = '';
		data.forEach(function(row) {
			html += '<tr>';
			html += '<td>' + (row.fee_number || '-') + '</td>';
			html += '<td>' + (row.student_name || '-') + '</td>';
			html += '<td>' + (row.student_number || '-') + '</td>';
			html += '<td>' + (row.payment_type || row.mode_of_payment || '-') + '</td>';
			html += '<td>' + format_currency(row.grand_total || 0) + '</td>';
			html += '<td>' + format_currency(row.receipt_amount || 0) + '</td>';
			html += '<td>' + format_currency(row.outstanding_amount || 0) + '</td>';
			html += '<td>' + (row.receipt_payment_date ? frappe.datetime.str_to_user(row.receipt_payment_date) : '-') + '</td>';
			html += '<td>' + (row.receipt_payer_name || '-') + '</td>';
			html += '<td><div class="action-buttons">';
			
			// View receipt button
			if (row.receipt_uploaded && row.attachments && row.attachments.length > 0) {
				html += '<button class="action-btn btn-view" onclick="fees_dashboard.view_receipt(\'' + row.fee_number + '\')">عرض الإيصال</button>';
			}
			
			// Accept receipt button
			if (row.receipt_uploaded && row.outstanding_amount > 0) {
				html += '<button class="action-btn btn-accept" onclick="fees_dashboard.accept_receipt(\'' + row.fee_number + '\', ' + (row.receipt_amount || row.outstanding_amount) + ')">قبول وتسديد</button>';
			}
			
			// Reject receipt button
			if (row.receipt_uploaded) {
				html += '<button class="action-btn btn-reject" onclick="fees_dashboard.reject_receipt(\'' + row.fee_number + '\')">رفض الإيصال</button>';
			}
			
			// View details button
			html += '<button class="action-btn btn-details" onclick="fees_dashboard.view_details(\'' + row.fee_number + '\')">التفاصيل</button>';
			
			html += '</div></td>';
			html += '</tr>';
		});
		
		tbody.html(html);
	};

	me.update_pagination = function(result) {
		let total = result.total_count || 0;
		let start = result.page_start || 0;
		let length = result.page_length || 20;
		let current_page = Math.floor(start / length) + 1;
		let total_pages = Math.ceil(total / length);
		
		// Update info
		$('#pagination-info').text('صفحة ' + current_page + ' من ' + total_pages + ' (' + total + ' سجل)');
		$('#table-info').text('إجمالي السجلات: ' + total);
		
		// Update buttons
		$('#btn-prev').prop('disabled', me.current_page === 0);
		$('#btn-next').prop('disabled', current_page >= total_pages);
		
		// Bind pagination events
		$('#btn-prev').off('click').on('click', function() {
			if (me.current_page > 0) {
				me.current_page--;
				me.load_data();
			}
		});
		
		$('#btn-next').off('click').on('click', function() {
			if (current_page < total_pages) {
				me.current_page++;
				me.load_data();
			}
		});
	};

	// Initialize
	me.init_filters();
	me.load_data();
	me.load_summary();
	
	// Store in global scope for button callbacks
	window.fees_dashboard = {
		view_receipt: function(fee_name) {
			frappe.call({
				method: 'education.education.page.fees_dashboard.fees_dashboard.get_fee_details',
				args: {
					fee_name: fee_name
				},
				callback: function(r) {
					if (r.message && !r.message.error) {
						let details = r.message;
						let attachments = details.attachments || [];
						
						if (attachments.length === 0) {
							frappe.msgprint(__('لا يوجد إيصال مرفق'));
							return;
						}
						
						let dialog = new frappe.ui.Dialog({
							title: __('عرض الإيصال - ' + fee_name),
							fields: [
								{
									fieldtype: 'HTML',
									fieldname: 'receipt_preview',
									options: '<div class="receipt-preview" id="receipt-preview"></div>'
								}
							],
							size: 'large'
						});
						
						let preview_html = '';
						attachments.forEach(function(att) {
							if (att.file_url) {
								let file_ext = att.file_name.split('.').pop().toLowerCase();
								if (['jpg', 'jpeg', 'png', 'gif', 'webp'].includes(file_ext)) {
									preview_html += '<img src="' + att.file_url + '" style="max-width: 100%; margin-bottom: 10px;"><br>';
								} else {
									preview_html += '<a href="' + att.file_url + '" target="_blank" class="btn btn-primary">تحميل ' + att.file_name + '</a><br>';
								}
							}
						});
						console.log(preview_html);
						dialog.fields_dict.receipt_preview.$wrapper.html(preview_html);
						dialog.show();
					} else {
						frappe.msgprint(__('حدث خطأ في تحميل الإيصال'));
					}
				}
			});
		},
		
		accept_receipt: function(fee_name, default_amount) {
			frappe.call({
				method: 'education.education.page.fees_dashboard.fees_dashboard.get_fee_details',
				args: {
					fee_name: fee_name
				},
				callback: function(r) {
					if (r.message && !r.message.error) {
						let details = r.message;
						
						let dialog = new frappe.ui.Dialog({
							title: __('قبول الإيصال وتسجيل الدفعة - ' + fee_name),
							fields: [
								{
									fieldtype: 'Section Break',
									label: 'معلومات الرسوم'
								},
								{
									fieldtype: 'Column Break'
								},
								{
									fieldname: 'student_name',
									fieldtype: 'Data',
									label: 'اسم الطالب',
									default: details.student_name,
									read_only: 1
								},
								{
									fieldname: 'outstanding_amount',
									fieldtype: 'Currency',
									label: 'المبلغ المستحق',
									default: details.outstanding_amount,
									read_only: 1
								},
								{
									fieldname: 'receipt_amount',
									fieldtype: 'Currency',
									label: 'مبلغ الإيصال',
									default: details.receipt_amount || default_amount,
									read_only: 1
								},
								{
									fieldtype: 'Section Break',
									label: 'تسجيل الدفعة'
								},
								{
									fieldname: 'payment_amount',
									fieldtype: 'Currency',
									label: 'مبلغ الدفعة',
									default: details.receipt_amount || default_amount,
									reqd: 1
								}
							],
							primary_action_label: __('تسجيل الدفعة'),
							primary_action: function(values) {
								let payment_amount = values.payment_amount;
								
								if (!payment_amount || payment_amount <= 0) {
									frappe.msgprint(__('يرجى إدخال مبلغ صحيح'));
									return;
								}
								
								frappe.call({
									method: 'education.education.page.fees_dashboard.fees_dashboard.accept_receipt_and_pay',
									args: {
										fee_name: fee_name,
										payment_amount: payment_amount
									},
									callback: function(r) {
										if (r.message && !r.message.error) {
											frappe.show_alert({
												message: __('تم تسجيل الدفعة بنجاح'),
												indicator: 'green'
											}, 5);
											dialog.hide();
											me.load_data();
											me.load_summary();
										} else {
											frappe.msgprint(r.message.error || __('حدث خطأ في تسجيل الدفعة'));
										}
									}
								});
							}
						});
						
						dialog.show();
					} else {
						frappe.msgprint(__('حدث خطأ في تحميل بيانات الرسوم'));
					}
				}
			});
		},
		
		reject_receipt: function(fee_name) {
			frappe.confirm(
				__('هل أنت متأكد من رفض هذا الإيصال؟'),
				function() {
					frappe.call({
						method: 'education.education.page.fees_dashboard.fees_dashboard.reject_receipt',
						args: {
							fee_name: fee_name
						},
						callback: function(r) {
							if (r.message && !r.message.error) {
								frappe.show_alert({
									message: __('تم رفض الإيصال بنجاح'),
									indicator: 'green'
								}, 5);
								me.load_data();
								me.load_summary();
							} else {
								frappe.msgprint(r.message.error || __('حدث خطأ في رفض الإيصال'));
							}
						}
					});
				}
			);
		},
		
		view_details: function(fee_name) {
			frappe.call({
				method: 'education.education.page.fees_dashboard.fees_dashboard.get_fee_details',
				args: {
					fee_name: fee_name
				},
				callback: function(r) {
					if (r.message && !r.message.error) {
						let details = r.message;
						
						let components_html = '';
						if (details.components && details.components.length > 0) {
							components_html = '<table class="components-table"><thead><tr><th>الفئة</th><th>الوصف</th><th>المبلغ</th><th>الخصم</th><th>الإجمالي</th></tr></thead><tbody>';
							details.components.forEach(function(comp) {
								components_html += '<tr>';
								components_html += '<td>' + (comp.fees_category || '-') + '</td>';
								components_html += '<td>' + (comp.description || '-') + '</td>';
								components_html += '<td>' + format_currency(comp.fee_rate || 0) + '</td>';
								components_html += '<td>' + (comp.discount || 0) + '%</td>';
								components_html += '<td>' + format_currency(comp.amount || 0) + '</td>';
								components_html += '</tr>';
							});
							components_html += '</tbody></table>';
						}
						
						let dialog = new frappe.ui.Dialog({
							title: __('تفاصيل الرسوم - ' + fee_name),
							fields: [
								{
									fieldtype: 'HTML',
									options: `
										<div class="fee-details-section">
											<div class="fee-details-grid">
												<div class="fee-detail-item">
													<div class="fee-detail-label">اسم الطالب</div>
													<div class="fee-detail-value">${details.student_name || '-'}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">تاريخ الإصدار</div>
													<div class="fee-detail-value">${details.posting_date ? frappe.datetime.str_to_user(details.posting_date) : '-'}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">تاريخ الاستحقاق</div>
													<div class="fee-detail-value">${details.due_date ? frappe.datetime.str_to_user(details.due_date) : '-'}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">الإجمالي</div>
													<div class="fee-detail-value">${format_currency(details.grand_total || 0)}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">المبلغ المدفوع</div>
													<div class="fee-detail-value">${format_currency(details.paid_amount || 0)}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">المبلغ المستحق</div>
													<div class="fee-detail-value">${format_currency(details.outstanding_amount || 0)}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">طريقة الدفع</div>
													<div class="fee-detail-value">${details.payment_type || details.mode_of_payment || '-'}</div>
												</div>
												<div class="fee-detail-item">
													<div class="fee-detail-label">حالة الإيصال</div>
													<div class="fee-detail-value">${details.receipt_uploaded ? 'تم الرفع' : 'لم يتم الرفع'}</div>
												</div>
											</div>
										</div>
										<div class="fee-details-section">
											<h5>مكونات الرسوم</h5>
											${components_html || '<p>لا توجد مكونات</p>'}
										</div>
									`
								}
							],
							size: 'large'
						});
						
						dialog.show();
					} else {
						frappe.msgprint(__('حدث خطأ في تحميل التفاصيل'));
					}
				}
			});
		}
	};
}

function format_currency(amount) {
	return frappe.format(amount);
}
